src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs

```diff
@@ -2977,8 +2977,30 @@ static void Main(string[] args)
                 //             case (object)"b":
                 Diagnostic(ErrorCode.ERR_ConstantExpected, @"(object)""b""").WithLocation(23, 18));
         }
-    }
 
+        [Fact]
+        [WorkItem(24869, "https://github.com/dotnet/roslyn/issues/24869")]
+        public void TestSelfReferencingViaLambda()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    static void F(IEnumerable<C> c)
+    {
+        const int F =
+        c.Select(o => new { E = F });
+    }
+}";
+            var comp = CreateStandardCompilation(source, references: new[] { LinqAssemblyRef });
+            comp.VerifyDiagnostics(
+                // (9,9): error CS0029: Cannot implicitly convert type 'System.Collections.Generic.IEnumerable<<anonymous type: int E>>' to 'int'
+                //         c.Select(o => new { E = F });
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "c.Select(o => new { E = F })").WithArguments("System.Collections.Generic.IEnumerable<<anonymous type: int E>>", "int").WithLocation(9, 9)
+                );
+        }
```

Comment:
This looks good, but... just curious... where did the "circular constant" diagnostic go? Did it get suppressed because of the error on its initializer?

Reply:
The CS0110 diagnostic gets stored in the local symbol's `_constantTuple` and would be reported once someone asks for `GetConstantValueDiagnostics`. That's done by `BindVariableDeclaration`, but only if there are no other diagnostics for the local:
```C#
...
       else if (kind == LocalDeclarationKind.Constant && initializerOpt != null && !localDiagnostics.HasAnyResolvedErrors())
            {
                var constantValueDiagnostics = localSymbol.GetConstantValueDiagnostics(initializerOpt);
...
```

Reply:
Can we test that case? Perhaps: `const int F = c.Sum(o => F);`